apps/rush-lib/src/logic/UpgradeRushSelf.ts

```diff
@@ -0,0 +1,259 @@
+    this._isDebug = options.isDebug;
+    this._terminal = new Terminal(
+      new ConsoleTerminalProvider({
+        verboseEnabled: options.isDebug
+      })
+    );
+  }
+
+  public getAvaiableRushVersions(): string[] {
+    this._terminal.writeLine('Fetching rush versions');
+
+    const { versions }: { versions: string[] } = this._npmView('@microsoft/rush');
+    if (!Array.isArray(versions)) {
+      throw new Error('Get @microsoft/rush versions incorrectly');
+    }
+
+    return versions;
+  }
+
+  public async upgrade(): Promise<IUpgradeResult> {
+    const versions: string[] = this.getAvaiableRushVersions();
+
+    const promptModule: inquirerTypes.PromptModule = inquirer.createPromptModule();
+    const { version } = await promptModule({
+      type: 'list',
+      name: 'version',
+      message: `Select a version to upgrade`,
+      choices: versions.sort((a, b) => semver.rcompare(a, b))
+    });
+
+    // update rush.json
+    const rushJsonFile: string = this._rushConfiguration.rushJsonFile;
+    const json: JsonObject = JsonFile.load(rushJsonFile);
+    if (json.rushVersion === version) {
+      console.log();
+      console.log(colors.yellow(`upgrading same rush version: ${version}`));
+      return {
+        needRushUpdate: false
+      };
+    }
+    json.rushVersion = version;
+    JsonFile.save(json, rushJsonFile, { updateExistingFile: true, onlyIfChanged: true });
+    if (this._isDebug) {
+      console.log(colors.gray(`upgrade rushVersion in rush.json to ${version}`));
+    }
+
+    const dependencies: Record<string, string> = this._npmView(`@microsoft/rush@${version}`, 'dependencies');
+    const targetPackageNames: string[] = Object.keys(dependencies).filter(
+      (k) => k.startsWith('@microsoft/') || k.startsWith('@rushstack/')
+    );
+    let needRushUpdate: boolean = false;
+    if (targetPackageNames.length) {
+      const targetPackages: Record<string, string> = {};
+      for (const packageName of targetPackageNames) {
+        targetPackages[packageName] = dependencies[packageName];
+      }
+
+      for (const project of this._rushConfiguration.projects) {
+        const packageJsonFile: string = path.resolve(project.projectFolder, 'package.json');
+        const packageJson: JsonObject = JsonFile.load(packageJsonFile);
+        for (const [k, v] of Object.entries(targetPackages)) {
+          if ('dependencies' in packageJson && k in packageJson.dependencies) {
+            packageJson.dependencies[k] = v;
+          }
+          if ('devDependencies' in packageJson && k in packageJson.devDependencies) {
+            packageJson.devDependencies[k] = v;
+          }
+        }
+
+        const saved: boolean = JsonFile.save(packageJson, packageJsonFile, {
+          updateExistingFile: true,
+          onlyIfChanged: true
+        });
+        if (saved) {
+          if (this._isDebug) {
+            console.log(colors.gray(`${project.packageName} package.json changed`));
```

Comment:
You can handle this by just calling `writeVerboseLine` on the `terminal` object.